src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -250,6 +285,277 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+        // No check required here: `permits_basic` checks that the issuer is non-empty
+        // and `ChainBuilder::potential_issuers` enforces subject/issuer matching,
+        // meaning that an CA with an empty subject cannot occur in a built chain.
+
+        let extensions = cert.extensions()?;
+
+        // NOTE: This conceptually belongs in `valid_issuer`, but is easier
+        // to test here. It's also conceptually an extension policy, but
+        // requires a bit of extra external state (`current_depth`) that isn't
+        // presently convenient to push into that layer.
+        if let Some(bc) = extensions.get_extension(&BASIC_CONSTRAINTS_OID) {
+            let bc: BasicConstraints = bc
+                .value()
+                .map_err(|_| PolicyError::Other("issuer has malformed basicConstraints"))?;
+
+            if bc
+                .path_length
+                .map_or(false, |len| current_depth as u64 > len)
+            {
+                return Err(PolicyError::Other("path length constraint violated"));
+            }
+        }
+
+        for ext_policy in self.ca_extension_policies.iter() {
+            ext_policy.permits(self, cert, &extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether the given EE certificate is compatible with this policy.
+    pub(crate) fn permits_ee(&self, cert: &Certificate<'_>) -> Result<(), PolicyError> {
+        self.permits_basic(cert)?;
+
+        let extensions = cert.extensions()?;
+        for ext_policy in self.ee_extension_policies.iter() {
+            ext_policy.permits(self, cert, &extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether `issuer` is a valid issuing CA for `child` at a
+    /// path-building depth of `current_depth`.
+    ///
+    /// This checks that `issuer` is permitted under this policy and that
+    /// it was used to sign for `child`.
+    ///
+    /// On success, this function returns the new path-building depth. This
+    /// may or may not be a higher number than the original depth, depending
+    /// on the kind of validation performed (e.g., whether the issuer was
+    /// self-issued).
+    pub(crate) fn valid_issuer(
+        &self,
+        issuer: &Certificate<'_>,
+        child: &Certificate<'_>,
+        current_depth: u8,
+    ) -> Result<u8, PolicyError> {
+        // The issuer needs to be a valid CA at the current depth.
+        self.permits_ca(issuer, current_depth)?;
+
+        let pk = self
+            .ops
+            .public_key(issuer)
+            .map_err(|_| PolicyError::Other("issuer has malformed public key"))?;
+        if self.ops.verify_signed_by(child, pk).is_err() {
+            return Err(PolicyError::Other("signature does not match"));
+        }
+
+        // Self-issued issuers don't increase the working depth.
+        match cert_is_self_issued(issuer) {
+            true => Ok(current_depth),
+            false => Ok(current_depth + 1),
```

Comment:
This can overflow :-)

Reply:
[90162e4](https://github.com/pyca/cryptography/pull/8873/commits/90162e436fd38dad887f597d5cdf31bc0489c727)